### PR TITLE
reserialize_bank_fields_with_hash

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem

We will be moving the accounts hash calculation from accounts_background_service to accounts_hash_verifier. This means we will not have an accounts hash until after we serialize the bank. But, we want to free the bank. So, we serialize the bank in accounts_background_service, allow the bank to be released, then the accounts_hash_verifier service thread picks up the AccountsPackage. It then runs hash calc to compute an accounts hash. Then, we can deserialize the bank, update the hash in the bank, and reserialize the bank. We then move on to build the snapshot.

Moving hash calc keeps it from blocking flush, clean, and shrink in accounts_background_service.

For reference, here is part of the prototype change to move the hash calc. I need to rebase it and clean it up.
#23689

#### Summary of Changes

Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
@mvines please look at the serialization/deserialization parts here.
@brooksprumo @yhchiang-sol @HaoranYi

**Comment by brooksprumo** [comment]
Could you add tests for this as well?

**Comment by mvines** [comment]
This feels clunky:
1. We write to disk, then read from disk, and write back to disk. I'd rather see a reworking of core/ to avoid the extra read and write 
2. Replacing the original file with a new file containing updated data seems like it may cause hard to debug issues: eg. somebody reads the original file before the hash is written into it and ends up with the wrong value.  If we can't solve (1), it'd be nice to at least use new filename for the bank-with-correct-hash data.

File: runtime/src/serde_snapshot/tests.rs [hunk 1]
```diff
@@ -214,6 +218,47 @@ fn test_bank_serialize_style(serde_style: SerdeStyle) {
     )
     .unwrap();
 
+    let accounts_hash = if update_accounts_hash {
```
Comment (jeffwashington) [inline runtime/src/serde_snapshot/tests.rs:221]: @brooksprumo here's a test. I understand more will change based on slack discussions, but this is at least the framework of a test that should exercise what we need.

File: runtime/src/serde_snapshot/tests.rs [hunk 2]
```diff
@@ -214,6 +218,52 @@ fn test_bank_serialize_style(serde_style: SerdeStyle) {
...
+        let post_path = slot_dir.join(slot.to_string());
+        let mut pre_path = post_path.clone();
+        pre_path.set_extension(BANK_SNAPSHOT_PRE_FILENAME_EXTENSION);
+        std::fs::create_dir(&slot_dir).unwrap();
+        {
+            let mut f = std::fs::File::create(pre_path.clone()).unwrap();
+            f.write_all(&buf).unwrap();
+        }
+        assert!(reserialize_bank_with_new_accounts_hash(
+            &temp_dir.path(),
```
Comment (brooksprumo) [inline runtime/src/serde_snapshot/tests.rs:244]: nit: Is the `&` needed here?

File: runtime/src/serde_snapshot/tests.rs [hunk 3]
```diff
@@ -250,6 +300,7 @@ fn test_bank_serialize_style(serde_style: SerdeStyle) {
     assert_eq!(dbank.get_balance(&key1.pubkey()), 0);
     assert_eq!(dbank.get_balance(&key2.pubkey()), 10);
     assert_eq!(dbank.get_balance(&key3.pubkey()), 0);
+    assert_eq!(dbank.get_accounts_hash(), accounts_hash);
```
Comment (brooksprumo) [inline runtime/src/serde_snapshot/tests.rs:303]: 👍

**Review by brooksprumo — COMMENTED** [review]
I think the overall design looks good. A few nits and questions below:

File: runtime/src/serde_snapshot.rs [hunk 4]
```diff
@@ -303,19 +315,55 @@ where
...
-        // atm, this just copies the file and deletes the old one
-        std::fs::copy(&bank_pre, bank_post).unwrap();
+    let mut found = false;
+    {
+        let file = std::fs::File::open(bank_pre.clone());
+        // some tests don't create the file
+        if let Ok(file) = file {
+            found = true;
+            let file_out = std::fs::File::create(bank_post).unwrap();
+            reserialize_bank_fields_with_new_hash(
```
[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/serde_snapshot.rs:355]: the code is setup to handle new and old serialization. I was trying to follow that model.

File: runtime/src/serde_snapshot/newer.rs [hunk 5]
```diff
@@ -230,7 +230,7 @@ impl<'a> TypeContext<'a> for Context {
                 )
             }));
         let slot = serializable_db.slot;
-        let hash = serializable_db
+        let hash: BankHashInfo = serializable_db
```
[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/serde_snapshot/newer.rs:233]: no, but it was helpful to know what was coming out of here.

**Review by brooksprumo — APPROVED** [review]
lgtm

**Comment by jeffwashington** [comment]
passed ci as part of this:
https://buildkite.com/solana-labs/solana/builds/69814#16e37342-840d-4822-9964-49f3f2fc1d88

**Pull request merged**